Allow accounting for direct-to-zone DOAS on zone sizing.

wfbuhl opened the pull request:
An addition to Sizing:Zone input allows the user to specify that a direct-to-zone Dedicated Outdoor Air System's heating or cooling effect needs to be added to the zone design heating or cooling loads. The user can specify DOAS supply temperatures and some limited information about the DOAS operation strategy.

wfbuhl committed:
Add new input items to idd file. Modify get routines and data structures to read and store data. Implement CalcDOASSupCondsForSizing.

wfbuhl committed:
Messed up but need to save

wfbuhl committed:
unit test of DOASEffectOnZoneSizing running. DOAS sizing code executing. Failing an array assertion test somewhere. Need to  commit code and check that my base version is working

wfbuhl committed:
cleared up array problem. New and old input files now run OK.

wfbuhl committed:
Put in the DOAS zone sizing variables and arrays. Current values are saved in SizeZoneEquipment. Still need to save the sequences.

wfbuhl committed:
Looks like it is working and reporting out DOAS effect on sizing om eio file. Need to add DOAS effect to predefined table reports.

wfbuhl committed:
Added MinOA to zone design peak cooling and heating tabular reports

wfbuhl committed:
Finally (!!) managed to get the thermostat setpoints to come out on the HVAC sizing tabular reports.

wfbuhl committed:
Add to the predefined reports

wfbuhl committed:
All reports printing

wfbuhl committed:
Debug reporting. Final unit test.

wfbuhl committed:
Deleted one file

wfbuhl committed:
Deleted one file

wfbuhl committed:
Merge branch 'develop' into 91958608DOASEffectOnZoneSizing

Conflicts:
	src/EnergyPlus/ZoneAirLoopEquipmentManager.cc
	tst/EnergyPlus/unit/CMakeLists.txt
	tst/EnergyPlus/unit/SQLite.unit.cc

wfbuhl committed:
Save changes to Engineering Reference and IOReference

wfbuhl committed:
Adding example file 5ZoneFanCoilDOASCool.idf. Also replacing ZoneTempPredictorCorrector.unit.cc which I somehow lost.

mbadams5 commented:
@wfbuhl There are still more changes to SQLite code needed for the new output field. That is why SQLiteFixture.addSQLiteZoneSizingRecord unit test is failing. I can expand on this comment later, or I may just go ahead and fix the necessary issues.

wfbuhl commented:
Hum - I can take a look at it tomorrow if you don't want to do it, though a
suggestion would always help. At some point I think I forgot about SQL and
only focused on my stuff.

Fred

On Mon, Jun 29, 2015 at 1:26 PM, Mark Adams notifications@github.com
wrote:

> @wfbuhl https://github.com/wfbuhl There are still more changes to
> SQLite code needed for the new output field. That is why
> SQLiteFixture.addSQLiteZoneSizingRecord unit test is failing. I can expand
> on this comment later, or I may just go ahead and fix the necessary issues.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5002#issuecomment-116832607.

mbadams5 committed:
Add DOASHeatAddRate to ZoneSizes table

wfbuhl committed:
Added DOAS row to Zone Component Load Summary Report. Redid Sensible Heat Gain Summary report so that rows sum to zero.

wfbuhl committed:
Final form of testfile for this branch

wfbuhl commented:
Because of changes to the IDD, the hourly report variables, and the eio the regression tests will fail for all example files. Additionally the predefined reports have changed so the html files will not be the same.

What should be the same are the csv outputs. I have checked some of these on my own machine and they are OK. I can't tell if they are OK in these regression tests but I think that is what they are saying. 

The predefined reports change are as follows:

HVAC Sizing Summary - Zone Cooling and Zone Heating. We've added 2 columns - "minimum outdoor airflow rate" and "heat gain from DOAS". Also "Thermostat setpoint at peak load' was coming out zero for duat setpoint cases and this has now been corrected.

Sensible Heat Gain Summary has been changed: the 1st 2 columns column  have been split into "HVAC Zone Eq & Other Sensible Air Heating" and "HVAC Terminal Unit Sensible Air Heating" and ditto for Cooling.

Lastly Zone Component Load Summary has had a row added: "Doas Direct to Zone".

Fred

Myoldmopar commented:
I think this fixes #3770.  @wfbuhl Can you confirm?

wfbuhl commented:
It certainly fixes this:

"If a DOAS is allowed to operate at cold or hot temperatures, and it is
sequenced #1 https://github.com/NREL/EnergyPlus/issues/1, then it can
easily happen that the equipment #2
https://github.com/NREL/EnergyPlus/issues/2 is not sized large enough to
handle the loads from the DOAS. "

It doesn't deal with all possible cases in which eq#1 could affect the size
of eq#2. But it should deal with most DOAS cases.

Fred

On Thu, Aug 13, 2015 at 7:50 AM, Edwin Lee notifications@github.com wrote:

> I think this fixes #3770 https://github.com/NREL/EnergyPlus/issues/3770.
> @wfbuhl https://github.com/wfbuhl Can you confirm?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5002#issuecomment-130717447.

wfbuhl committed:
Merge branch 'develop' into 91958608DOASEffectOnZoneSizing

Conflicts:
	src/EnergyPlus/OutputReportPredefined.cc
	src/EnergyPlus/ZoneEquipmentManager.cc
	tst/EnergyPlus/unit/CMakeLists.txt

wfbuhl committed:
Mistakenly deleted 811 and 812 IDD files. Just putting them back.

Myoldmopar commented:
@wfbuhl Whenever you get around to fixing the failing build above, make sure to pull in the latest develop so there aren't unrelated diffs from CI.

wfbuhl committed:
Merge branch 'develop' into 91958608DOASEffectOnZoneSizing

wfbuhl commented:
I'm back and just doing it now.

On Tue, Aug 18, 2015 at 8:24 AM, Edwin Lee notifications@github.com wrote:

> @wfbuhl https://github.com/wfbuhl Whenever you get around to fixing the
> failing build above, make sure to pull in the latest develop so there
> aren't unrelated diffs from CI.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5002#issuecomment-132249123.

wfbuhl committed:
Fixed double declaration of AirDistUnit in ZoneTempPredictorCorrector.cc

wfbuhl committed:
Eliminate Linux Build Warnings. Fix array bounds bomb in several inputs.

wfbuhl committed:
Fix more load warnings. Fix unit test.

wfbuhl commented:
So all build warnings are eliminated. Integration and unit tests are OK. The fails are expected but I will double check them. Looks like this is just about ready to go.

wfbuhl commented:
This is ready to merge

wfbuhl committed:
Merge branch 'develop' into 91958608DOASEffectOnZoneSizing

Myoldmopar committed:
Merge remote-tracking branch 'origin/develop' into 91958608DOASEffectOnZoneSizing

Myoldmopar commented:
@wfbuhl CI went kinda insane and ~~left you lots of comments~~.  I'm getting the CI machines restarted and cleaned up.  In light of this, I pulled develop into this branch as well and we should get some once-and-for-all good CI results.  If all is well, I'll review and merge.

Update: I cleaned up lots of the unnecessary comments.

Myoldmopar committed:
Cleanup member constructors, comments, other minor changes

Myoldmopar commented:
@wfbuhl @mjwitte OK, I looked over the source and made some changes that shouldn't affect functionality at all, just cleanup.  I looked at the CI results and confirmed that the changes are only as expected.  Once CI verifies I didn't break anything with my minor commit, I'm OK with this going in.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -30616,9 +30616,33 @@ Sizing:Zone,
+  A7, \field Design Specification Zone Air Distribution Object Name
       \type object-list
       \object-list DesignSpecificationZoneAirDistributionNames
+  A8, \field Account for Dedicated Outside Air System
```
@wfbuhl Outside should be Outdoor everywhere for consistency with other field names.

mjwitte commented on idd/Energy+.idd.in:
```diff
@@ -30616,9 +30616,33 @@ Sizing:Zone,
+      \key NeutralDehumidifiedSupplyAir
+      \key ColdSupplyAir
+      \default NeutralSupplyAir
+  N17,\field Dedicated Outside Air Low Setpoint for Design
```
@wfbuhl This field and the next should say . . . Temperature Setpoint . . .

Myoldmopar commented:
@wfbuhl Could you provide a "synopsis for users" as the title of this PR, and also maybe edit the first comment to describe the changes in a couple sentences or bullet points?  That way if the user clicks on it in the Changelog and they make it to this page, they'll be able to quickly see the changes brought in by this branch.

Thanks!

wfbuhl committed:
Incorporate Mike's suggestions. Namely replace "outside" with "outdoor" in the IDD and docs. Add "temperature" before "setpoint".

Myoldmopar committed:
Fix line endings in IO-01b

Myoldmopar commented:
@wfbuhl @mjwitte I fixed a line ending issue in the docs and pushed that up.  I was about to say this is ready to go in, but I noticed there are diffs in _SmallOffice_Dulles.  Are they expected?  From a quick glance they might be small, but not completely insignificant.

wfbuhl commented:
Let me take a look.

Fred

On Thu, Sep 3, 2015 at 5:14 AM, Edwin Lee notifications@github.com wrote:

> @wfbuhl https://github.com/wfbuhl @mjwitte https://github.com/mjwitte
> I fixed a line ending issue in the docs and pushed that up. I was about to
> say this is ready to go in, but I noticed there are diffs in
> _SmallOffice_Dulles. Are they expected? From a quick glance they might be
> small, but not completely insignificant.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/NREL/EnergyPlus/pull/5002#issuecomment-137419265.

wfbuhl commented:
I believe it is OK. I changed where AirTerminal:SingleDuct:Uncontrolled().HeatRate etc. is calculated to make it consistent with the other terminal units and with SumMCpDTsystem. This is to get the Sensible Heat Gain Components report to work correctly (the rows must sum to zero) and get my new column to fit in with how the other columns are calculated. Now this stuff is calculated in ZoneTempPredictorCorrector:

for ( NodeNum = 1; NodeNum <= ZoneEquipConfig( ZoneEquipConfigNum ).NumInletNodes; ++NodeNum ) {
				// Get node conditions
				NodeTemp = Node( ZoneEquipConfig( ZoneEquipConfigNum ).InletNode( NodeNum ) ).Temp;
				MassFlowRate = Node( ZoneEquipConfig( ZoneEquipConfigNum ).InletNode( NodeNum ) ).MassFlowRate;
				CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );

				SumMCpDTsystem += MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );

			} // NodeNum

			if ( ZoneEquipConfig( ZoneEquipConfigNum ).ADUNum > 0 ) {
				ADUNum = ZoneEquipConfig( ZoneEquipConfigNum ).ADUNum;
				NodeTemp = Node( AirDistUnit( ADUNum ).OutletNodeNum ).Temp;
				MassFlowRate = Node( AirDistUnit( ADUNum ).OutletNodeNum ).MassFlowRate;
				CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );
				ADUHeatAddRate = MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );
				AirDistUnit( ADUNum ).HeatRate = max( 0.0, ADUHeatAddRate );
				AirDistUnit( ADUNum ).CoolRate = std::abs( min( 0.0, ADUHeatAddRate ) );
				AirDistUnit( ADUNum ).HeatGain = AirDistUnit( ADUNum ).HeatRate * TimeStepSys * SecInHour;
				AirDistUnit( ADUNum ).CoolGain = AirDistUnit( ADUNum ).CoolRate * TimeStepSys * SecInHour;
			}

			if ( ZoneEquipConfig( ZoneEquipConfigNum ).SDUNum > 0 ) {
				SDUNum = ZoneEquipConfig( ZoneEquipConfigNum ).SDUNum;
				NodeTemp = Node( DirectAir( SDUNum ).ZoneSupplyAirNode ).Temp;
				MassFlowRate = Node( DirectAir( SDUNum ).ZoneSupplyAirNode ).MassFlowRate;
				CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );
				SDUHeatAddRate = MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );
				DirectAir( SDUNum ).HeatRate = max( SDUHeatAddRate, 0.0 );
				DirectAir( SDUNum ).CoolRate = std::abs( min( SDUHeatAddRate, 0.0 ) );
				DirectAir( SDUNum ).HeatEnergy = DirectAir( SDUNum ).HeatRate * TimeStepSys * SecInHour;
				DirectAir( SDUNum ).CoolEnergy = DirectAir( SDUNum ).CoolRate * TimeStepSys * SecInHour;
			}




:

wfbuhl commented:
I believe it is OK. I moved where the output variables of AirTerminal:SingleDuct:Uncontrolled are calculated from DirectAirManager to ZoneTempPredictorCorrector so that they are calculated identically with SumMCpDTsystem. Also moved AirDistUnit().HeatRate etc. to ZoneTempPredictorCorrector. All this to make the Sensible Heat Gain Summary report come out correctly: the columns have to sum to zero and I've split an existing column into two columns. Here is where this stuff is calculated now in ZoneTempPredictorCorrector starting at line 4936:

```
        for ( NodeNum = 1; NodeNum <= ZoneEquipConfig( ZoneEquipConfigNum ).NumInletNodes; ++NodeNum ) {
            // Get node conditions
            NodeTemp = Node( ZoneEquipConfig( ZoneEquipConfigNum ).InletNode( NodeNum ) ).Temp;
            MassFlowRate = Node( ZoneEquipConfig( ZoneEquipConfigNum ).InletNode( NodeNum ) ).MassFlowRate;
            CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );

            SumMCpDTsystem += MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );

        } // NodeNum

        if ( ZoneEquipConfig( ZoneEquipConfigNum ).ADUNum > 0 ) {
            ADUNum = ZoneEquipConfig( ZoneEquipConfigNum ).ADUNum;
            NodeTemp = Node( AirDistUnit( ADUNum ).OutletNodeNum ).Temp;
            MassFlowRate = Node( AirDistUnit( ADUNum ).OutletNodeNum ).MassFlowRate;
            CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );
            ADUHeatAddRate = MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );
            AirDistUnit( ADUNum ).HeatRate = max( 0.0, ADUHeatAddRate );
            AirDistUnit( ADUNum ).CoolRate = std::abs( min( 0.0, ADUHeatAddRate ) );
            AirDistUnit( ADUNum ).HeatGain = AirDistUnit( ADUNum ).HeatRate * TimeStepSys * SecInHour;
            AirDistUnit( ADUNum ).CoolGain = AirDistUnit( ADUNum ).CoolRate * TimeStepSys * SecInHour;
        }

        if ( ZoneEquipConfig( ZoneEquipConfigNum ).SDUNum > 0 ) {
            SDUNum = ZoneEquipConfig( ZoneEquipConfigNum ).SDUNum;
            NodeTemp = Node( DirectAir( SDUNum ).ZoneSupplyAirNode ).Temp;
            MassFlowRate = Node( DirectAir( SDUNum ).ZoneSupplyAirNode ).MassFlowRate;
            CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );
            SDUHeatAddRate = MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );
            DirectAir( SDUNum ).HeatRate = max( SDUHeatAddRate, 0.0 );
            DirectAir( SDUNum ).CoolRate = std::abs( min( SDUHeatAddRate, 0.0 ) );
            DirectAir( SDUNum ).HeatEnergy = DirectAir( SDUNum ).HeatRate * TimeStepSys * SecInHour;
            DirectAir( SDUNum ).CoolEnergy = DirectAir( SDUNum ).CoolRate * TimeStepSys * SecInHour;
        }
```

You can also see the code change on GitHub. The reason for the difference is Cp is different and the zone temperature is slightly different.

Myoldmopar commented:
Thanks for looking into it @wfbuhl, your justification sounds satisfactory and thus your branch wins and will go in first.  This evening I will try to help folks get all the other PR branches updated with this latest develop so that their branches don't show diffs.

Merging in exactly five minutes :clock5: :clock4: :clock3: :clock2: :clock1: :clock12: (_those aren't minutes..._), unless someone gives me a reason not to, @mjwitte.
